pkg/hs/natpmp-static/hsrc_lib/Network/NATPMP.hsc

```diff
@@ -0,0 +1,260 @@
+          <*> (#{peek natpmpresp_t, pnu.newportmapping.lifetime} p)
+      RTTCPPortMapping ->
+        NatPmpResponseTCPPortMapping
+          <$> (#{peek natpmpresp_t, pnu.newportmapping.privateport} p)
+          <*> (#{peek natpmpresp_t, pnu.newportmapping.mappedpublicport} p)
+          <*> (#{peek natpmpresp_t, pnu.newportmapping.lifetime} p)
+
+  poke _ _ = error "Responses are an output data structure; poke makes no sense"
+
+type NatPmpResponseHandle = Ptr NatPmpResponse
+
+foreign import ccall unsafe "getgateway.h getdefaultgateway" _get_default_gateway :: Ptr CUInt -> IO CInt
+
+foreign import ccall unsafe "natpmp.h initnatpmp" _init_nat_pmp :: NatPmpHandle -> CInt -> CInt -> IO CInt
+foreign import ccall unsafe "natpmp.h closenatpmp" _close_nat_pmp :: NatPmpHandle -> IO CInt
+foreign import ccall unsafe "natpmp.h sendpublicaddressrequest" sendPublicAddressRequest :: NatPmpHandle -> IO CInt
+foreign import ccall unsafe "natpmp.h sendnewportmappingrequest" sendNewPortMappingRequest :: NatPmpHandle -> CInt -> CUShort -> CUShort -> CUInt -> IO CInt
+
+foreign import ccall unsafe "binding.h readNatResponseSynchronously" readNatResponseSynchronously :: NatPmpHandle -> NatPmpResponseHandle -> IO CInt
+
+-- Give the type system some help
+_peekCUInt :: Ptr CUInt -> IO CUInt
+_peekCUInt = peek
+
+uintToEnum :: Enum e => CUInt -> e
+uintToEnum = toEnum . fromIntegral
+
+intToEnum :: Enum e => CInt -> e
+intToEnum = toEnum . fromIntegral
+
+
+-- Fetches the default gateway as an ipv4 address
+getDefaultGateway :: IO (Maybe HostAddress)
+getDefaultGateway =
+  alloca $ \(pReturnAddr :: Ptr CUInt) -> do
+      ret <- _get_default_gateway pReturnAddr
+      case ret of
+        0 -> (Just . fromIntegral) <$> _peekCUInt pReturnAddr
+        _ -> pure Nothing
+
+
+data RespType
+  = RTPublicAddress
+  | RTUDPPortMapping
+  | RTTCPPortMapping
+  deriving (Eq, Show)
+
+instance Enum RespType where
+  fromEnum RTPublicAddress = 0
+  fromEnum RTUDPPortMapping = 1
+  fromEnum RTTCPPortMapping = 2
+
+  toEnum 0 = RTPublicAddress
+  toEnum 1 = RTUDPPortMapping
+  toEnum 2 = RTTCPPortMapping
+  toEnum unmatched = error ("RespType.toEnum: Cannot match " ++ show unmatched)
+
+
+data ProtocolType
+  = PTUDP
+  | PTTCP
+  deriving (Eq, Show)
+
+instance Enum ProtocolType where
+  fromEnum PTUDP = 1
+  fromEnum PTTCP = 2
+
+  toEnum 1 = PTUDP
+  toEnum 2 = PTTCP
+  toEnum x = error ("ProtocolType.toEnum: Cannot match " ++ show x)
+
+
+data Error
+  = ErrInvalidArgs
+  | ErrSocketError
+  | ErrCannotGetGateway
+  | ErrCloseErr
+  | ErrRecvFrom
+  | ErrNoPendingReq
+  | ErrNoGatewaySupport
+  | ErrConnectErr
+  | ErrWrongPacketSource
+  | ErrSendErr
+  | ErrFcntlError
+  | ErrGetTimeOfDayError
+  --
+  | ErrUnsuportedVersion
+  | ErrUnsupportedOpcode
+  --
+  | ErrUndefinedError
+  | ErrNotAuthorized
+  | ErrNetworkFailure
+  | ErrOutOfResources
+  --
+  | ErrTryAgain
+  | ErrHaskellBindingsErr
+  deriving (Eq, Show)
+
+instance Enum Error where
```

Comment:
I'm not. I'm using hsc2hs, and my understanding is that its #enum is meant to be used to emit a series of #const statements. (And we're using hsc2hs here because we already use it in lmdb.)